### PR TITLE
Addition of ItemMergeEvent BUKKIT-1959

### PR DESCRIPTION
**Pull request by deathmarine**
https://github.com/Bukkit/CraftBukkit/pull/1001 Adjacent.

### CODE REVIEW

**Comment by Wolvereness** [comment]
This needs clear documentation of which entity will cease to exist, if either.

On another, DISCUSSION note, what if the event was also considered cancelled if the entities would no longer properly merge? As this is a "state" kind of situation, where a plugin might change the internal stack for either entity, I see unanticipated functionality if the event continued after they would no longer properly merge (proactively cancel it). My thoughts (slightly pseudo):

```
isCancelled() { return this.cancelled = this.cancelled || !entity1.item.isSimilar(entity2.item); }
```

**Comment by deathmarine** [comment]
Note the position in nms in which each entity is taken from, it just grabs them both prior to its checks. If the Item(Entity) is changed on either side it will follow the original checks for sake of normality. (Trying to forcibly merge a diamond axe with a block of wool will do nothing but leave two entity items there.)

I'll update the documentation to be more direct which Item(entity) initiated the merge attempt. From nms the "merged" entity will always cease to exist, if the merge can be completed.

SideNote:  I see some events are static and some are instantiated.
EntityItem - PlayerPickupItemEvent & ItemDespawnEvent what would you consider the standard.

**Comment by EdGruberman** [comment]
@deathmarine "From nms the "merged" entity will always cease to exist, if the merge can be completed."
That "if" is a big if.  Based on event processing, it can not be determined beforehand which item will cease to exist.  Safer to just say neither is guaranteed.

**Comment by riking** [comment]
See my comment on Bukkit/CraftBukkit#1001, this event needs to be rewritten.

File: src/main/java/org/bukkit/event/entity/ItemMergeEvent.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    public void setCancelled(boolean cancel) {
+        cancelled = cancel;
+    }
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/ItemMergeEvent.java:28]: Extra spaces

File: src/main/java/org/bukkit/event/entity/ItemMergeEvent.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Returns the entity involved with the merge.
+     * 
+     * @return Item
+     */
+    @Override
+    public Item getEntity() {
+        return (Item) entity;
+    }
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/ItemMergeEvent.java:38]: Extra spaces

File: src/main/java/org/bukkit/event/entity/ItemMergeEvent.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    
+    /**
+     * Returns the merged entity involved with the event.
+     * 
+     * @return Item
+     */
+    public Item getMerged() {
+        return (Item) item;
+    }
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/ItemMergeEvent.java:47]: Extra spaces.

**Comment by turt2live** [comment]
@deathmarine If you are still interested in this please update it to the latest code base and address the concerns mentioned above. The PR will also need updating to comply with the CONTRIBUTING.md guidelines. 

Please also provide a decent justification for why this PR should be used over Ed's (GitHub autolinked them above for you).

**Comment by turt2live** [comment]
Closed due to lack of activity and for not following the contributing guidelines after well over a year.

**Pull request closed without merging**